src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2StreamWindowManager.cs

```diff
@@ -0,0 +1,257 @@
+
+            public void AdjustWindow(int bytesConsumed)
+            {
+                Debug.Assert(bytesConsumed > 0);
+                Debug.Assert(_deliveredBytes < StreamWindowThreshold);
+
+                if (!_stream.ExpectResponseData)
+                {
+                    // We are not expecting any more data (because we've either completed or aborted).
+                    // So no need to send any more WINDOW_UPDATEs.
+                    return;
+                }
+
+                if (_windowScalingEnabled)
+                {
+                    AdjustWindowDynamic(bytesConsumed);
+                }
+                else
+                {
+                    AjdustWindowStatic(bytesConsumed);
+                }
+            }
+
+            private void AjdustWindowStatic(int bytesConsumed)
+            {
+                _deliveredBytes += bytesConsumed;
+                if (_deliveredBytes < StreamWindowThreshold)
+                {
+                    return;
+                }
+
+                int windowUpdateIncrement = _deliveredBytes;
+                _deliveredBytes = 0;
+
+                Task sendWindowUpdateTask = _connection.SendWindowUpdateAsync(_stream.StreamId, windowUpdateIncrement);
+                _connection.LogExceptions(sendWindowUpdateTask);
+            }
+
+            private void AdjustWindowDynamic(int bytesConsumed)
+            {
+                _deliveredBytes += bytesConsumed;
+
+                if (_deliveredBytes < StreamWindowThreshold)
+                {
+                    return;
+                }
+
+                int windowUpdateIncrement = _deliveredBytes;
+                long currentTime = Stopwatch.GetTimestamp();
+
+                if (_connection._rttEstimator.MinRtt > TimeSpan.Zero)
+                {
+                    TimeSpan rtt = _connection._rttEstimator.MinRtt;
+                    TimeSpan dt = StopwatchTicksToTimeSpan(currentTime - _lastWindowUpdate);
+
+                    // We are detecting bursts in the amount of data consumed within a single 'dt' window update period.
+                    // The value "_deliveredBytes / dt" correlates with the bandwidth of the connection.
+                    // We need to extend the window, if the bandwidth-delay product grows over the current window size.
+                    // To enable empirical fine tuning, we apply a configurable multiplier (_windowScaleThresholdMultiplier) to the window size, which defaults to 1.0
+                    //
+                    // The condition to extend the window is:
+                    // (_deliveredBytes / dt) * rtt > _streamWindowSize * _windowScaleThresholdMultiplier
+                    //
+                    // Which is reordered into the form below, to avoid the division:
+                    if (_deliveredBytes * rtt.Ticks > _streamWindowSize * dt.Ticks * _windowScaleThresholdMultiplier)
```

Comment:
Is there any risk of overflow here?

Reply:
With extremely high RTT, `_deliveredBytes * rtt.Ticks` can theoretically overflow `long`. Converting to `double` before the multiplication should mitigate this.